### PR TITLE
Change function constructors to WithFields

### PR DESCRIPTION
**Pull request by electriclilies**
In this PR, I change function constructors to WithFields. I tried to change all of them but may have missed a few.

### CODE REVIEW

**Comment by electriclilies** [comment]
i split these out from another branch, there are still some residual changes from that branch that I need to revert.

**Comment by electriclilies** [comment]
This is ready for review, please take a look @mbs-octoml @mikepapadim. 
Note that I rebased against #9826, so you'll see some changes from that in the diff until it is merged

File: src/relay/transforms/to_cps.cc [hunk 1]
```diff
@@ -311,15 +313,17 @@ Function ToCPS(const Function& f, const IRModule& m) {
 Function UnCPS(const Function& f) {
   CheckFeature(f, FeatureSet::All() - fGraph);
   ICHECK_GT(f->params.size(), 0);
-  std::vector<Var> new_params;
+  Array<Var> new_params;
   for (const auto& p : f->params) {
-    new_params.push_back(Var(p->name_hint(), p->checked_type()));
+    // TODO(@electriclilies): Not sure if this is correct, it was copying before,
+    // but seems like we just need to make a copy to pop so should be fine?
+    new_params.push_back(WithFields(std::move(p)));
```
Comment (mbs-octoml) [inline src/relay/transforms/to_cps.cc:320]: I'd retain the copies here -- it is building a new function with new parameters.

I think including all args including the span in the ctor will ensure any new fields will at least trigger a compilation error so these few examples of explicit shallow copies can be maintained. Adding a generic ShallowCopy or Clone helper seems overkill.

**Review by mbs-octoml — APPROVED** [review]
Thanks Lily! Just two minor changes asking to retain the existing shallow copies. I think this still needs to be rebased but no mind.

File: src/relay/transforms/inline.cc [hunk 2]
```diff
@@ -131,7 +130,9 @@ class Inliner : ExprMutator {
     const auto* fn = base_func.as<FunctionNode>();
     ICHECK(fn) << "Expected to work on a Relay function.";
 
-    auto func = Function(fn->params, fn->body, fn->ret_type, fn->type_params, fn->attrs);
+    // TODO(@electriclilies): If Function is a COW node, then if it gets written to we shouldn't
+    // have any sharing, right? So we don't need to reconstruct?
+    Function func = WithFields(GetRef<Function>(fn));
```
[the remark this answers is not in the document]
Comment (electriclilies) [reply src/relay/transforms/inline.cc]: Done!

File: src/relay/transforms/to_cps.cc (hunk 1 above)
Comment (electriclilies) [reply src/relay/transforms/to_cps.cc]: Done, thanks!

**Review by jroesch — APPROVED** [review]

**Pull request merged**